File: articles.jdbc/src/test/java/pet/db/jdbc/tool/generator/ArticleTestDataGenerator.java
package pet.db.jdbc.tool.generator;

import lombok.RequiredArgsConstructor;

import org.instancio.Instancio;
import org.instancio.Select;

import org.springframework.stereotype.Component;

import pet.db.jdbc.model.dto.payload.NewArticlePayload;
import pet.db.jdbc.model.dto.Article;
import pet.db.jdbc.model.dto.User;
import pet.db.jdbc.service.ArticleService;

import java.time.LocalDateTime;
import java.util.List;

@Component
@RequiredArgsConstructor
public class ArticleTestDataGenerator implements TestDataGenerator<Article> {

    private final TestDataGenerator<User> userTestDataGenerator;

    private final ArticleService articleService;

    @Override
    public Article generateSavedData() {
        return generateSavedData(1).getFirst();
    }

    @Override
    public Article generateUnsavedData() {
        return generateUnsavedData(1).getFirst();
    }

    @Override
    public List<Article> generateSavedData(Integer dataSize) {
        return generateUnsavedData(dataSize).stream()
                .map(article -> articleService.create(article, generateSavedAuthorIds(dataSize)))
                .toList();
    }

    @Override
    public List<Article> generateUnsavedData(Integer dataSize) {
        return Instancio
                .ofList(Article.class)
                .size(dataSize)
                .set(Select.field(Article::getDateOfCreation), LocalDateTime.now())
                .generate(Select.field(Article::getTopic), gen -> gen.string()
                        .length(NewArticlePayload.TOPIC_MIN_LENGTH, NewArticlePayload.TOPIC_MAX_LENGTH)
                        .alphaNumeric())
                .generate(Select.field(Article::getContent), gen -> gen.string()
                        .length(NewArticlePayload.CONTENT_MIN_LENGTH, NewArticlePayload.CONTENT_MAX_LENGTH)
                        .alphaNumeric())
                .create();
    }

    private List<Integer> generateSavedAuthorIds(int dataSize) {
        return userTestDataGenerator.generateSavedData(dataSize)
                .stream()
                .map(User::getId)
                .toList();
    }
}

File: src/test/kotlin/pet/articles/tool/generator/ArticleTestDataGenerator.kt
package pet.articles.tool.generator

import net.datafaker.Faker

import pet.articles.model.dto.User
import pet.articles.service.article.ArticleService
import pet.articles.model.dto.Article
import pet.articles.tool.extension.generateRandom
import pet.articles.web.validation.ArticlePayloadValidation.Fields.Constraints.CONTENT_MAX_LENGTH
import pet.articles.web.validation.ArticlePayloadValidation.Fields.Constraints.TOPIC_MAX_LENGTH

import java.time.LocalDateTime

class ArticleTestDataGenerator(
    private val articleService: ArticleService,
    private val userGenerator: TestDataGenerator<User>
) : TestDataGeneratorBaseImpl<Article>(
    generate = ::generate,
    create = { article ->
        articleService.create(article, generateSavedAuthorIds(userGenerator))
    },
    toInvalidState = ::makeInvalid
) {

    companion object {
        private const val ARTICLE_FIELD_TOPIC_INVALID_LENGTH = 1000
        private const val CONTENT_PARAGRAPH_SIZE = 3
        private const val CONTENT_SENTENCE_SIZE = 3
        private const val MIN_NUM_OF_TEST_AUTHOR_IDS = 1
        private const val MAX_NUM_OF_TEST_AUTHOR_IDS = 5

        private fun generate(faker: Faker = Faker()) = Article(
            id = faker.number().positive(),
            dateOfCreation = LocalDateTime.now(),
            topic = faker
                .lorem()
                .sentence(CONTENT_SENTENCE_SIZE)
                .take(TOPIC_MAX_LENGTH),
            content = faker
                .lorem()
                .paragraph(CONTENT_PARAGRAPH_SIZE)
                .take(CONTENT_MAX_LENGTH)
        )

        private fun makeInvalid(article: Article) = article.copy(
            topic = String.generateRandom(ARTICLE_FIELD_TOPIC_INVALID_LENGTH)
        )

        fun generateSavedAuthorIds(
            userGenerator: TestDataGenerator<User>
        ): List<Int> {
            val numOfAuthors: Int = (MIN_NUM_OF_TEST_AUTHOR_IDS..MAX_NUM_OF_TEST_AUTHOR_IDS)
                .random()
            return userGenerator.generateSavedData(numOfAuthors)
                .map { it.id!! }
        }
    }
}